src/sql/SqlResult.ts

```diff
@@ -0,0 +1,360 @@
+            this.closeDeferred.resolve();
+        }).catch(this.closeDeferred.reject);
+
+        return this.closeDeferred.promise
+    }
+
+    /** Called when next page of the result is received. */
+    onNextPage(page: SqlPage) {
+        this.currentPage = page;
+        this.currentRowCount = page.getRowCount();
+        this.currentPosition = 0;
+
+        if (page.isLast()) {
+            this.last = true;
+            this.closed = true;
+        }
+    }
+
+    /** Called when an error is occurred during SQL execute */
+    onExecuteError(error: HazelcastSqlException): void {
+        if (this.closed) return;
+        this.updateCount = Long.fromInt(-1);
+        this.rowMetadata = null;
+        this.executeDeferred.reject(error);
+    }
+
+    /** Returns the current row. Used by {@link next}. */
+    getCurrentRow(): SqlRowType {
+        if (this.returnRawResult) { // Return SqlRow
+            const values = [];
+            for (let i = 0; i < this.currentPage.getColumnCount(); i++) {
+                values.push({
+                    name: this.rowMetadata.getColumns()[i].name,
+                    value: this.serializationService.toObject(this.currentPage.getValue(this.currentPosition, i))
+                });
+            }
+            return new SqlRowImpl(values, this.rowMetadata);
+        } else { // Return objects
+            const result: SqlRowAsObject = {};
+            for (let i = 0; i < this.currentPage.getColumnCount(); i++) {
+                const columnMetadata = this.rowMetadata.getColumnByIndex(i);
+                if (columnMetadata != null) {
+                    result[columnMetadata.name] = this.serializationService.toObject(
+                        this.currentPage.getValue(this.currentPosition, i)
+                    );
+                }
+            }
+            return result;
+        }
+    }
+
+    /** Called when a execute response is received. */
+    onExecuteResponse(rowMetadata: SqlRowMetadata | null, rowPage: SqlPage, updateCount: Long) {
+        // Ignore the response if SQL result is closed.
+        if (this.closed) return;
+
+        if (rowMetadata !== null) { // Result that including rows
+            this.rowMetadata = rowMetadata;
+            this.onNextPage(rowPage);
+            this.updateCount = Long.fromInt(-1);
+        } else { // Result that including update count
+            this.updateCount = updateCount;
+            this.closed = true;
+        }
+        this.executeDeferred.resolve(true);
+    }
+
+    /**
+     * Fetches the next page. Called internally by iteration logic. Pages are fetched whenever the current page is fully
+     * iterated.
+     */
+    fetch(): Promise<SqlPage> {
+        // If there is an ongoing fetch, return that promise
+        if (this.fetchDeferred?.promise)
+            return this.fetchDeferred.promise;
+
+        // Do not start a fetch if close is called
+        if (this.closeDeferred?.promise) {
```

Comment:
Are you sure? Java side sets an error to the fetch result. 

```java
try {
    // Do nothing if the result is already closed.
    if (closed) {
        return;
    }

    // If the cancellation is initiated before the first response is received, then throw cancellation errors on
    // the dependent methods (update count, row metadata, iterator).
    if (state == null) {
        onExecuteError(QueryException.cancelledByUser());
    }

    // Make sure that all subsequent fetches will fail.
    if (fetch == null) {
        fetch = new SqlFetchResult();
    }

    onFetchFinished(null, QueryException.cancelledByUser());

    // Send the close request.
    service.close(connection, queryId);
} finally {
    // Set the closed flag to avoid multiple close requests.
    closed = true;
}
```

with the line 

```java
onFetchFinished(null, QueryException.cancelledByUser());
```

subsequent calls to fetch causes the same error to be thrown again.

Reply:
Ok, you are right, I guess I mis-remembered it